cartofacade/carto_facade_queue.go

```diff
@@ -0,0 +1,151 @@
+	Position
+	// InternalState can be used to represent the viam_carto_get_internal_state call in c.
+	InternalState
+	// PointCloudMap can be used to represent the viam_carto_get_point_cloud_map in c.
+	PointCloudMap
+)
+
+// InputType defines the type being provided as input to the work.
+type InputType int64
+
+const (
+	// Reading represents a lidar reading input into c funcs.
+	Reading InputType = iota
+	// Timestamp represents the timestamp input into c funcs.
+	Timestamp
+)
+
+// WorkItem defines one piece of work that can be put on the queue.
+type WorkItem struct {
+	Result   chan interface{}
+	workType WorkType
+	inputs   map[InputType]interface{}
+}
+
+// DoWork provides the logic to call the correct cgo functions with the correct input.
+func (w *WorkItem) DoWork(
+	q *Queue,
+) (interface{}, error) {
+	switch w.workType {
+	case Initialize:
+		return New(q.CartoConfig, q.CartoAlgoConfig, *q.CartoLib)
+	case Start:
+		return nil, q.Carto.Start()
+	case Stop:
+		return nil, q.Carto.Stop()
+	case Terminate:
+		return q.Carto.Terminate(), nil
+	case AddSensorReading:
+		reading, ok := w.inputs[Reading].([]byte)
+		if !ok {
+			return nil, errors.New("could not cast inputted byte to byte slice")
+		}
+
+		timestamp, ok := w.inputs[Timestamp].(time.Time)
+		if !ok {
+			return nil, errors.New("could not cast inputted timestamp to times.Time")
+		}
+
+		return nil, q.Carto.AddSensorReading(reading, timestamp)
+	case Position:
+		return q.Carto.GetPosition()
+	case InternalState:
+		return q.Carto.GetInternalState()
+	case PointCloudMap:
+		return q.Carto.GetPointCloudMap()
+	}
+	return nil, fmt.Errorf("no worktype found for: %v", w.workType)
+}
+
+// Queue represents a queue to consume work from and enforce one call into C at a time.
+type Queue struct {
+	WorkChannel     chan WorkItem
+	CartoLib        *CartoLib
+	Carto           Carto
+	CartoConfig     CartoConfig
+	CartoAlgoConfig CartoAlgoConfig
+}
+
+// NewQueue instantiates the Queue struct.
+func NewQueue(cartoLib *CartoLib, cartoCfg CartoConfig, cartoAlgoCfg CartoAlgoConfig) Queue {
+	return Queue{
+		WorkChannel:     make(chan WorkItem),
+		Carto:           Carto{},
+		CartoLib:        cartoLib,
+		CartoConfig:     cartoCfg,
+		CartoAlgoConfig: cartoAlgoCfg,
+	}
+}
+
+// HandleIncomingRequest puts incoming requests on the queue and consumes from queue.
+func (q *Queue) HandleIncomingRequest(ctx context.Context, workType WorkType, inputs map[InputType]interface{}) interface{} {
+	// Question: What is reasonable timeout period?
+	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
+	defer cancel()
+
+	work := WorkItem{
+		Result:   make(chan interface{}),
+		workType: workType,
+		inputs:   inputs,
+	}
+
+	// wait until work can get put on the queue (and timeout if needed)
+	select {
+	case q.WorkChannel <- work:
+		// wait until result is put on result queue (and timeout if needed)
+		result := <-work.Result
+		return result
+	case <-ctx.Done():
```

Comment:
Can we please have a test for what happens to the work channel go routine if this timeout is hit?